### ISSUE TITLE
A Question on the Temperature in the Contrastive Loss

### ISSUE THREAD

**Issue by yaoyiran** [drboog/Lafite#25]
Hi Yufan, thank you again for your great work! I just got a small question on the role of temperature and find that there seems to be a small mismatch between the formula in the paper and the code and got a bit confused. It would be super helpful if you could kindly provide more details/thoughts.

In the paper the contrastive loss is defined in formula (6) and (7) as L = - &tau; log(A/B), but in the code [L153-L60](https://github.com/drboog/Lafite/blob/main/training/loss.py#L153-L160) in loss.py, it is actually L = - log(&tau; * A/B).  Considering that - log(&tau; * A/B) = - log(&tau;) - log(A/B), and the fact that the term - log(&tau;) is a constant, it will have no influence on the gradient. So I would recommend rewriting formula (6) and (7) in the paper as simply L = - log(A/B) without a &tau; term as a multiplier. Please correct me if my understanding is not correct.

I was also wondering if you could lmk why in your original design, there is a &tau; multiplier outside of the log() stuff, since in a standard InfoNCE loss, &tau; usually only appears in the exp(sth./&tau;) but is not used outside of the log() stuff.

**Comment by drboog** [comment]
Hi, you may want to check a previous discussion about this.
https://github.com/drboog/Lafite/issues/4
Basically, one of my previous version of code used this "incorrect" implementation, and I tuned hyper-parameters based on it. I corrected the implementation later, but didn't re-search the hyper-parameters for new implementation. 
After discussing with others that try to reproduce our results, I kept this implementation in the code, because it can easily reproduce our results in the paper. And I mentioned about this in the Appendix of our paper.
Personally, I think using "correct" implementation may lead to better results, but it requires some hyper-parameter tuning.

**Comment by yaoyiran** [comment]
@drboog Thanks for your message. But my question is different from the previous discussion #4. I know the "exponential sharpening" discussion in #4, and already saw that in your paper's appendix. And in your code, the exponential sharpening is only related to [L149](https://github.com/drboog/Lafite/blob/main/training/loss.py#L149) in loss.py and I already saw your comment there in the code.


However, my question is about the [L153-154](https://github.com/drboog/Lafite/blob/main/training/loss.py#L153-L154) and the &tau; outside of the log() in the equation below. I believe it is different from #4. Would you mind having a look at my comment above again? Thanks! 
<img width="330" alt="Screenshot 2022-11-02 at 23 29 18" src="https://user-images.githubusercontent.com/25559589/199621824-32ce7918-3658-4d73-a251-e67fc1f540bd.png">

**Comment by drboog** [comment]
Oh, I see. 

Yes, I think you are correct. That's actually a typo. I was trying to write that outside the log.
It doesn't influence the method, because there will be other hyper-parameters such as --itd, --itc outside the log. Now our itd just represent (itd*\tau). They are all just hyper-parameters to be tuned.
Thanks for point that out, I didn't notice that typo previously.

The reason why I write a \tau outside log is that this follows some previous papers. Basically, if you take the gradient w.r.t the loss, some \tau from the exponential terms will be cancelled. But as you can see, it doesn't really influence anything, I'm just used to writing in that way.
